libraries/AP_HAL_Linux/UARTDevice.cpp

```diff
@@ -25,99 +26,91 @@ UARTDevice::UARTDevice(const char *device_path):
 
 UARTDevice::~UARTDevice()
 {
-
+    close();
 }
 
 bool UARTDevice::close()
 {
-    if (_fd != -1) {
-        if (::close(_fd) < 0) {
-            return false;
-        }
-    }
-
-    _fd = -1;
+    delete _pollable;
+    _pollable = nullptr;
 
     return true;
 }
 
 bool UARTDevice::open()
 {
-    _fd = ::open(_device_path, O_RDWR | O_CLOEXEC);
+    int fd = ::open(_device_path, O_RDWR | O_CLOEXEC);
 
-    if (_fd < 0) {
+    if (fd < 0) {
         ::fprintf(stderr, "Failed to open UART device %s - %s\n",
                   _device_path, strerror(errno));
         return false;
     }
 
     _disable_crlf();
+    _pollable = new BufferedPollable(fd);
+
+    auto scheduler = Scheduler::from(hal.scheduler);
+    scheduler->register_pollable(_pollable, Poller::Event::All);
 
     return true;
 }
 
 ssize_t UARTDevice::read(uint8_t *buf, uint16_t n)
 {
-    return ::read(_fd, buf, n);
+    if (!_pollable) {
+        return -EAGAIN;
+    }
+
+    return _pollable->buffered_read(buf, n);
 }
 
 ssize_t UARTDevice::write(const uint8_t *buf, uint16_t n)
 {
-    struct pollfd fds;
-    fds.fd = _fd;
-    fds.events = POLLOUT;
-    fds.revents = 0;
-
-    int ret = 0;
-
-    if (poll(&fds, 1, 0) == 1) {
-        ret = ::write(_fd, buf, n);
+    if (!_pollable) {
+        return -EAGAIN;
     }
 
-    return ret;
+    return _pollable->buffered_write(buf, n);
 }
 
-void UARTDevice::set_blocking(bool blocking)
+void UARTDevice::set_blocking_mode(SerialDevice::BlockingMode mode)
 {
-    int flags = fcntl(_fd, F_GETFL, 0);
-    
-    if (blocking) {
-        flags = flags & ~O_NONBLOCK;
-    } else {
-        flags = flags | O_NONBLOCK;
+    if (!_pollable) {
+        errno = EAGAIN;
+        return;
     }
 
-    if (fcntl(_fd, F_SETFL, flags) < 0) {
-        ::fprintf(stderr, "Failed to make UART nonblocking %s - %s\n",
-                  _device_path, strerror(errno));
-    }
+    _pollable->set_blocking(mode & SerialDevice::BlockingMode::NonblockingIO);
 
+    BufferedPollable *bp = _pollable;
```

Comment:
Didn't understand this. Why not: _polable->set_blocking_writes and _pollable->set_blocking_reads?